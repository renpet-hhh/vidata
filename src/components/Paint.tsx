import React, { CSSProperties, useRef, useState, useEffect, Ref, useImperativeHandle } from 'react'
import useStack from '../hooks/useStack';
import CanvasContext from '../utils/CanvasContext';

export interface PaintMethods {
    getImageData: () => ImageData,
    undo: () => void,
    redo: () => void
}

interface Props {
    width: number,
    height: number,
    initialImageData?: ImageData,
    style?: CSSProperties,
    radius?: number,
    color?: string,
    mode?: "erase" | "getColor" | "normal",
    listeners?: {
        getColor: (color: string) => void
    }
    // styled-components
    className?: string
};

const DEFAULTS : {mode: "normal", radius: 5, color: "black"} = {
    mode: "normal",
    radius: 5,
    color: "black"
}


/** A customizable canvas wrapper that draws upon mouse drags and clicks, with a drawing history 
 * 
 * You have access to imperative functions by `ref`:
 * ```
 * {
 *      getImageData, // get current imageData
 *      undo, // redraws to the state before the last drawing
 *      redo // redraws to the state before the last undo
 * }
 * ```
 * 
 * And access to other actions by `listeners`:
 * ```
 * {
 *      getColor // called when user clicks at Paint while 'mode' is 'getColor'
 * }
 * ```
 * 
 * @example
 * const Parent = () => {
 *      const paintRef = useRef<PaintMethods>(null);
 *      const undo = () => paintRef.current && paintRef.current.undo();
 *      const getColorListener = (color: string) => console.log("user selected " + color);
 *      return (
 *          <div>
 *              <Paint ref={paintRef} width={200} height={200} listeners={{getColor: getColorListener}}></Paint>
 *              <button onClick={undo}>Undo</button>
 *          </div>
 *      );
 * };
*/
const Paint = React.forwardRef((props: Props, ref : Ref<PaintMethods>) => {
    if (props.style && props.style.padding !== undefined) console.warn("Paint doesn't support padding. Drawing will be out of place");

    const canvasRef = useRef<HTMLCanvasElement>(null);
    const [isDragging, setIsDragging] = useState(false);
    /** Stack for REDO */
    const historyTrash = useStack<ImageData | undefined>();
    /** Stack for UNDO */
    const history = useStack<ImageData | undefined>();
    /** Current canvas image data
     * (canvas getImageData and putImageData in a stateful approach) */
    const [imageData, setImageData] = useState<ImageData>();
    /** Blank ImageData */
    const blank = useRef<ImageData>();
    const CanvasCTXRef = useRef<CanvasContext>();

    const mode = props.mode !== undefined ? props.mode : DEFAULTS.mode;
    const radius = props.radius !== undefined ? props.radius : DEFAULTS.radius
    const color = props.color !== undefined ? props.color : DEFAULTS.color;

    const undo = () => {
        if (history.length() == 0) return;
        const last = history.pop();
        historyTrash.push(imageData);
        setImageData(last);
    }

    const redo = () => {
        if (historyTrash.length() == 0) return;
        const last = historyTrash.pop();
        history.push(imageData);
        setImageData(last);
    }

    const getCtxWrapper = () => {
        if (!CanvasCTXRef.current) {
            CanvasCTXRef.current = new CanvasContext(canvasRef.current!.getContext("2d")!, color, radius);
        }
        return CanvasCTXRef.current.update(color, radius);
    }

    /** Always use the `imageData` state hook when possible */
    const _getImageData = () => {
        if (canvasRef.current) {
            return canvasRef.current.getContext("2d")!.getImageData(0, 0, props.width, props.height);
        }
        throw new Error("Tried to get ImageData from an unmounted canvas");
    }

    /** Use initial canvas in browser to get blank ImageData (constructor not supported in NodeJS) */
    useEffect(() => {
        if (canvasRef.current) {
            blank.current = canvasRef.current.getContext("2d")!.getImageData(0, 0, props.width, props.height);
        }
    }, []);

    /** Enables canvas stateful approach.
     * Every time `imageData` changes, the canvas is redrawn
     */
    useEffect(() => {
        if (!canvasRef.current) return;
        if (!blank.current) {
            console.warn("Couldn't get the blank ImageData, useEffect order might be wrong");
            return;
        }
        if (imageData === undefined) {
            canvasRef.current.getContext("2d")!.putImageData(props.initialImageData || blank.current, 0, 0);
            return;
        }
        canvasRef.current.getContext("2d")!.putImageData(imageData, 0, 0);
    }, [imageData]);

    /** Send imperative methods to parent */
    useImperativeHandle(ref, () => ({
        getImageData: (() => {
            if (!blank.current) {
                throw new Error("Couldn't get the blank ImageData, useEffect order might be wrong");
            }
            return imageData || props.initialImageData || blank.current;
        }),
        undo, redo
    }), [imageData])


    const getClickPosition = (e: React.MouseEvent<HTMLCanvasElement>) => {
        if (!canvasRef.current) throw new Error("Tried to get click position in non existent canvas element");
        const RECT = canvasRef.current.getBoundingClientRect();
        const x = e.clientX - RECT.left;
        const y = e.clientY - RECT.top;
        return { x, y };
    }
    const onMouseDown = (e: React.MouseEvent<HTMLCanvasElement>) => {
        if (mode === "normal" || mode === "erase") {
            history.push(imageData);
            historyTrash.reset();
            const { x, y } = getClickPosition(e);
            const ctxWrapper = getCtxWrapper();
            mode === "erase" ? ctxWrapper.eraseCircle(x, y) : ctxWrapper.drawCircle(x, y);
            setIsDragging(true);
        }
    }
    const onMouseMove = !isDragging ? undefined : (e: React.MouseEvent<HTMLCanvasElement>) => {
        if (mode === "normal") {
            const { x, y } = getClickPosition(e);
            const ctxWrapper = getCtxWrapper();
            ctxWrapper.drawLineTo(x, y);
            ctxWrapper.drawCircle(x, y);
        } else if (mode === "erase") {
            const { x, y } = getClickPosition(e);
            const ctxWrapper = getCtxWrapper();
            ctxWrapper.eraseLineTo(x, y);
            ctxWrapper.eraseCircle(x, y);
        }
    }
    const onClick = (e: React.MouseEvent<HTMLCanvasElement>) => {
        if (mode === "getColor") {
            const { x, y } = getClickPosition(e);
            const ctxWrapper = getCtxWrapper();
            const color = ctxWrapper.getPixelColor(x, y);
            props.listeners && props.listeners.getColor(color);
        }
    }
    const cancelDrag = () => {
        if (mode === "normal" || mode === "erase") {
            if (!isDragging) return;
            setIsDragging(false);
            setImageData(_getImageData());
            const ctxWrapper = getCtxWrapper();
            ctxWrapper.resetPosition();
        }
    }
    return (
        <canvas ref={canvasRef} onMouseDown={onMouseDown} onMouseUp={cancelDrag} onMouseLeave={cancelDrag}
            onMouseMove={onMouseMove} onClick={onClick}
            style={props.style} width={props.width} height={props.height} className={props.className}
                data-testid="Paint"></canvas>
    );
});



export default Paint;