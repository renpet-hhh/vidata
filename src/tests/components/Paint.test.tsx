import React, { RefObject } from 'react';
import Paint, { PaintMethods } from '../../components/Paint';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { act } from '@testing-library/react';


afterEach(cleanup);

it('renders blank', () => {
    const paintRef : RefObject<PaintMethods> = {current: null};
    render(<Paint ref={paintRef} width={10} height={10}></Paint>);
    const imageData = paintRef.current!.getImageData();
    expect(imageData).toMatchSnapshot("should be blank");
});

it('renders initial ImageData', () => {
    const paintRef : RefObject<PaintMethods> = {current: null};
    const initialImageData = new ImageData(10, 10);
    initialImageData.data.set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    render(<Paint ref={paintRef} initialImageData={initialImageData} width={10} height={10}></Paint>);
    const imageData = paintRef.current!.getImageData();
    expect(imageData).toMatchSnapshot("should have [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] as first data");
});

it('draws on mousedown', async () => {
    const { getByTestId } = render(<Paint width={10} height={10}></Paint>);
    act(() => {
        fireEvent.mouseDown(getByTestId("Paint"), {clientX: 5, clientY: 5});
    });
    // @ts-ignore
    const ctx : CanvasRenderingContext2D = getByTestId("Paint").getContext("2d");
    expect(ctx.__getDrawCalls()).toMatchSnapshot("should draw arc at (5, 5)");
});

it('draws on drag', async () => {
    const { getByTestId } = render(<Paint width={10} height={10}></Paint>);
    act(() => {
        fireEvent.mouseDown(getByTestId("Paint"), {clientX: 5, clientY: 5});
    });
    act(() => {
        fireEvent.mouseMove(getByTestId("Paint"), {clientX: 6, clientY: 6});
    })
    // @ts-ignore
    const ctx : CanvasRenderingContext2D = getByTestId("Paint").getContext("2d");
    expect(ctx.__getDrawCalls()).toMatchSnapshot("should draw arc at (5, 5) and (6, 6)");
});

it(`doesn't draw on mousemove when not dragging`, async () => {
    const { getByTestId } = render(<Paint width={10} height={10}></Paint>);
    act(() => {
        fireEvent.mouseDown(getByTestId("Paint"), {clientX: 5, clientY: 5});
    });
    act(() => {
        fireEvent.mouseUp(getByTestId("Paint"));
    });
    act(() => {
        fireEvent.mouseMove(getByTestId("Paint"), {clientX: 6, clientY: 6});
    })
    // @ts-ignore
    const ctx : CanvasRenderingContext2D = getByTestId("Paint").getContext("2d");
    expect(ctx.__getDrawCalls()).toMatchSnapshot("should draw arc at (5, 5)");
});

it('draws correct color', () => {
    const { getByTestId } = render(<Paint width={10} height={10} color="#123456"></Paint>);
    act(() => {
        fireEvent.mouseDown(getByTestId("Paint"), {clientX: 2, clientY: 2});
    });
    // @ts-ignore
    const ctx : CanvasRenderingContext2D = getByTestId("Paint").getContext("2d");
    expect(ctx.__getEvents()).toMatchSnapshot("should have #123456 color");
});

it('draws correct radius', () => {
    const { getByTestId } = render(<Paint width={10} height={10} radius={4}></Paint>);
    act(() => {
        fireEvent.mouseDown(getByTestId("Paint"), {clientX: 2, clientY: 2});
    });
    // @ts-ignore
    const ctx : CanvasRenderingContext2D = getByTestId("Paint").getContext("2d");
    expect(ctx.__getDrawCalls()).toMatchSnapshot("should have radius 4");
});

it('erases', () => {
    const { getByTestId, rerender } = render(<Paint width={10} height={10} color="#123456"></Paint>);
    act(() => {
        fireEvent.mouseDown(getByTestId("Paint"), {clientX: 2, clientY: 2});
    });
    rerender(<Paint width={10} height={10} color="#123456" mode="erase"></Paint>);
    act(() => {
        fireEvent.mouseMove(getByTestId("Paint"), {clientX: 3, clientY: 3});
    });
    // @ts-ignore
    const ctx : CanvasRenderingContext2D = getByTestId("Paint").getContext("2d");
    expect(ctx.__getEvents()).toMatchSnapshot("should have erased at (3, 3)");
});

it('gets color in empty canvas', () => {
    // can't test for actual behavior
    const getColorListener = jest.fn();
    const { getByTestId } = render(<Paint width={10} height={10} color="#234567" mode="getColor" listeners={{getColor: getColorListener}}></Paint>);
    act(() => {
        fireEvent.click(getByTestId("Paint"), {clientX: 5, clientY: 5});
    });
    expect(getColorListener).toHaveBeenCalledTimes(1);
    expect(getColorListener).toHaveBeenLastCalledWith("rgba(0, 0, 0, 0)");
});
