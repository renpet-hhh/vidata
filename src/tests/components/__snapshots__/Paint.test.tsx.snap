// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't draw on mousemove when not dragging: should draw arc at (5, 5) 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 5,
            "y": 5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`draws correct color: should have #123456 color 1`] = `
Array [
  Object {
    "props": Object {
      "dirtyHeight": NaN,
      "dirtyWidth": NaN,
      "dirtyX": NaN,
      "dirtyY": NaN,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "putImageData",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#123456",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 5,
      "startAngle": 0,
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`draws correct radius: should have radius 4 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 4,
            "startAngle": 0,
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`draws on drag: should draw arc at (5, 5) and (6, 6) 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 5,
            "y": 5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 5,
            "y": 5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 6,
            "y": 6,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 6,
            "y": 6,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`draws on mousedown: should draw arc at (5, 5) 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 5,
            "y": 5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`erases: should have erased at (3, 3) 1`] = `
Array [
  Object {
    "props": Object {
      "dirtyHeight": NaN,
      "dirtyWidth": NaN,
      "dirtyX": NaN,
      "dirtyY": NaN,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "putImageData",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#123456",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 5,
      "startAngle": 0,
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "destination-out",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalCompositeOperation",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#123456",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 3,
      "y": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 3,
            "y": 3,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "value": "source-over",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalCompositeOperation",
  },
  Object {
    "props": Object {
      "value": "destination-out",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalCompositeOperation",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#123456",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 5,
      "startAngle": 0,
      "x": 3,
      "y": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 5,
            "startAngle": 0,
            "x": 3,
            "y": 3,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "source-over",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalCompositeOperation",
  },
]
`;

exports[`renders blank: should be blank 1`] = `
ImageData {
  "_data": Uint8ClampedArray [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "_height": 10,
  "_width": 10,
}
`;

exports[`renders initial ImageData: should have [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] as first data 1`] = `
ImageData {
  "_data": Uint8ClampedArray [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "_height": 10,
  "_width": 10,
}
`;
